### PR TITLE
[FEATURE] Allow customization of the items shown in browser

### PR DESCRIPTION
**Pull request by PeterPetrik**
Allow customization of the items shown in browser. User can decide (in the Interface Customization dialog) to hide some of the root items in the browser panel (e.g. Favourites, or POSTGIS provider, ...)

Screenshot 2020-01-09 at 09 17 05

Funded by Limerick City and County Council

### CODE REVIEW

**Comment by nyalldawson** [comment]
Is the intention here just to hide the items? If so, I think this should be done in the browser filter model...

**Comment by nirvn** [comment]
I was going to suggest that too. It'd be great to have a "hide item" right click menu item for browser items, and not only for root nodes.

**Comment by PeterPetrik** [comment]
No, intention is that the council wants to customize the browser widget, so the items are not visible when QGIS starts. Similar to other customizations. Difference between this and "hide" is that the items are not loaded at all at the first place when customization is used

**Comment by nyalldawson** [comment]
But if they are hidden the assumption is that they're unused... So I can't really see the benefit of not creating them but hiding them in the proxy

**Comment by wonder-sk** [comment]
The idea was to have some stuff from browser removed together with other parts of customization of the GUI. So it is not meant as a 'regular' feature for users to right click an item in browser and hide it through a menu...

We have discussed the approach with Peter earlier and played with various options including use of proxy model to hide things in browser. But this approach of not even creating items made more sense - it goes more in line with other features of customization where some GUI elements are removed (with no way of going back) and it is lighter on resources (not even create items that would be removed afterwards).

**Comment by nyalldawson** [comment]
> The idea was to have some stuff from browser removed together with other parts of customization of the GUI. So it is not meant as a 'regular' feature for users to right click an item in browser and hide it through a menu...

I definitely see the strong use case here -- I'm going to use this to immediately hide oracle & mssql nodes from some customer's deployments.

> We have discussed the approach with Peter earlier and played with various options including use of proxy model to hide things in browser. But this approach of not even creating items made more sense - it goes more in line with other features of customization where some GUI elements are removed (with no way of going back) and it is lighter on resources (not even create items that would be removed afterwards).

My concern with the "never create" vs "hide" approach is that other code may be expecting these root items to exist so we run the risk of crashes or unexpected behavior. I've seen MANY issues caused by users using the customization tool to remove bits of the QGIS ui which just breaks all kinds of assumptions in the QGIS code, resulting in a huge mess. I think the never-create approach is likely to lead to these same kind of issues. I'm also aware of plugins which access these root connection items (inasafe does, for example).

I don't think there'd be a significant expense incurred with creating these items vs hiding them. Creating the empty root connection nodes is very inexpensive, basically just a one-time creation of a QgsDataItem followed by an attempt to iterate through an empty QSettings connection group...

**Comment by nyalldawson** [comment]
(and code complexity should be very similar -- there's already QgsBrowserProxyModel which can be extended to handle this)

**Comment by PeterPetrik** [comment]
ok, going to refactor: `QgsDataItem` will have member variable provider key and in the `QgsBrowserModel` I  add new role to `ItemDataRole` that will be used in proxy model

File: python/core/auto_generated/qgsdataitem.sip.in [hunk 1]
```diff
@@ -300,6 +301,27 @@ Sets the ``name`` of the item (the displayed text for the item).
 
     QString path() const;
     void setPath( const QString &path );
+
+    QString providerKey() const;
+%Docstring
+Returns the provider key that created this item
+
+If key has a prefix "special:", it marks that the item was not created with a provider,
+but manually. For example "special: Favorites"
```
Comment (nyalldawson) [inline python/core/auto_generated/qgsdataitem.sip.in]: is the space between ":" and "Favorites" intentional here?

File: python/core/auto_generated/qgsbrowserproxymodel.sip.in [hunk 2]
```diff
@@ -131,6 +131,21 @@ filterByLayerType() is ``True``.
...
 
 .. seealso:: :py:func:`setFilterByLayerType`
+%End
+
+    void setDataItemProviderKeyFilter( const QStringList &filter );
+%Docstring
+Sets the customization filters for data items based on item's data provider key
+
+By default browser model shows all items from all available data items provider and few special
+items (e.g. Favourites). To customize the behavious, set the filter to not load certain data items.
```
Comment (nyalldawson) [inline python/core/auto_generated/qgsbrowserproxymodel.sip.in]:
```suggestion
items (e.g. Favourites). To customize the behaviour, set the filter to not load certain data items.
```

File: python/core/auto_generated/qgsbrowserproxymodel.sip.in [hunk 3]
```diff
@@ -131,6 +131,21 @@ filterByLayerType() is ``True``.
...
+The items that are not based on data item providers (e.g. Favourites, Home) have
+prefix "special:".
+
```
Comment (nyalldawson) [inline python/core/auto_generated/qgsbrowserproxymodel.sip.in]: I'd suggest adding some specific examples here - eg. the string for postgres and the string for the favourites or home item -- it's not immediately clear from the docs alone.

File: python/gui/auto_generated/qgsbrowserdockwidget.sip.in [hunk 4]
```diff
@@ -58,6 +58,20 @@ Returns the message bar associated with the dock.
...
 
 .. versionadded:: 3.6
+%End
+
+    void setDisabledDataItemsKeys( const QStringList &filter );
+%Docstring
+Sets the customization for data items based on item's data provider key
+
+By default browser model shows all items from all available data items provider and few special
+items (e.g. Favourites). To customize the behavious, set the filter to not load certain data items.
```
Comment (nyalldawson) [inline python/gui/auto_generated/qgsbrowserdockwidget.sip.in]:
```suggestion
items (e.g. Favourites). To customize the behaviour, set the filter to not load certain data items.
```

File: src/app/qgscustomization.cpp [hunk 5]
```diff
@@ -696,6 +700,42 @@ void QgsCustomization::createTreeItemStatus()
...
+  items << QStringList( {QStringLiteral( "special:Volumes" ), tr( "Volume Folder (MacOS only)" )} );
+
+  const auto constProviders = QgsApplication::dataItemProviderRegistry()->providers();
+  for ( QgsDataItemProvider *pr : constProviders )
+  {
+    int capabilities = pr->capabilities();
+    if ( capabilities != QgsDataProvider::NoDataCapabilities )
+    {
+      QStringList item;
+      item << QStringLiteral( "%1" ).arg( pr->name() ) << QStringLiteral( "Data Item Provider: %1" ).arg( pr->name() );
```
Comment (nyalldawson) [inline src/app/qgscustomization.cpp]:
```suggestion
      item << QStringLiteral( "%1" ).arg( pr->name() ) << QObject::tr( "Data Item Provider: %1" ).arg( pr->name() );
```

File: src/core/qgsbrowserproxymodel.cpp [hunk 6]
```diff
@@ -230,3 +233,35 @@ bool QgsBrowserProxyModel::filterAcceptsItem( const QModelIndex &sourceIndex ) c
...
+bool QgsBrowserProxyModel::filterAcceptsProviderKey( const QModelIndex &sourceIndex ) const
+{
+  if ( !mModel )
+    return true;
+
+  const QString providerKey = mModel->data( sourceIndex, QgsBrowserModel::ProviderKeyRole ).toString();
+  if ( providerKey.isEmpty() )
+    return true;
+
+  return ( !mHiddenDataItemsKeys.contains( providerKey ) );
```
Comment (nyalldawson) [inline src/core/qgsbrowserproxymodel.cpp]:
```suggestion
  return !mHiddenDataItemsKeys.contains( providerKey );
```

**Review by nyalldawson — COMMENTED** [review]
Looking good -- I'm much happier with this approach, thanks!

Can you extend the tests as https://github.com/qgis/QGIS/blob/master/tests/src/core/testqgsbrowserproxymodel.cpp and https://github.com/qgis/QGIS/blob/master/tests/src/core/testqgsbrowsermodel.cpp to test the provider key role and filtering?

**Comment by PeterPetrik** [comment]
would like to merge tomorrow (wed)

File: python/gui/auto_generated/qgsbrowserdockwidget.sip.in (hunk 4 above)
Comment (nyalldawson) [reply python/gui/auto_generated/qgsbrowserdockwidget.sip.in]: Fix typo

**Review by nyalldawson — COMMENTED** [review]
Looks good!

**Comment by nyalldawson** [comment]
Just need to fix the "behavious" typos

**Comment by elpaso** [comment]
@PeterPetrik nice API, I'm using it right now but I've found the naming a bit misleading:

1. the documentation mentions "Sets a filter to hide data items based on on item's data provider key" but `filterAcceptsProviderKey` does not actually receive a provider key but an item name, which is a complete different thing (for instance it receives `GPKG` that is **not** a data provider key, or `PostGIS` instead `postgres`)

2. "special:Favourites" is mispelled

3. it is not immediately clear that a filter is a blacklist (my first thought before reading the documentation was that it was a white list)

I would recommend to rename the method to something that does not mention data provider keys (but maybe data item names) and make it clear the it is an **exclusion** (blacklist) filter and not a white list.

**Comment by PeterPetrik** [comment]
Hi @elpaso ,

just finally got to your points, sorry for late reply. Looks like you addressed most of it in https://github.com/qgis/QGIS/commit/31aeb6e351fb710da5338db2c2166d10c5cca415 , thank you very much for cleaning that part for me.

adressed 2. in https://github.com/qgis/QGIS/pull/37849

**Comment by elpaso** [comment]
@PeterPetrik thanks! It's funny that I've sent you an email 30 minutes ago before reading this message, telepathy?

**Pull request merged**